client_apis/python/src/tests/integration/user_test.py

```diff
@@ -0,0 +1,252 @@
+            }
+            return user
+
+        @classmethod
+        def gen_global_admin(cls):
+            user = cls.gen_user()
+            user['global_admin'] = True
+
+            return user
+
+    ## List user tests
+
+    def test_list_users(self):
+        users = cb.user_enum()
+        self.assertIsNotNone(users)
+        self.assertGreater(len(users), 0)
+
+        for user in users:
+            self._verify_retrieved_user(user)
+
+    ## Get user tests
+
+    def test_get_current_user(self):
+        cur_user = cb.user_info(None)
+        self.assertIsNotNone(cur_user)
+
+        self._verify_retrieved_user(cur_user)
+
+    ## Create user tests
+
+    def test_create_user_no_username(self):
+        user_data = self.TestDataGen.gen_user()
+
+        user_add_params = self._convert_user_data_to_user_add_params(user_data)
+        user_add_params['username'] = None
+
+        # should not be able to create user with no username
+        with self.assertRaises(requests.HTTPError) as cm:
+            cb.user_add_from_data(**user_add_params)
+
+        self.assertEqual(cm.exception.response.status_code, 400)
+
+    def test_create_user_bad_password_confirm(self):
+        user_data = self.TestDataGen.gen_user()
+
+        user_add_params = self._convert_user_data_to_user_add_params(user_data)
+        user_add_params['confirm_password'] = "wr0ngp@ssw0rd"
+
+        # should not be able to create user with bad password confirmation
+        with self.assertRaises(requests.HTTPError) as cm:
+            cb.user_add_from_data(**user_add_params)
+
+        self.assertEqual(cm.exception.response.status_code, 400)
+
+    def test_create_valid_user(self):
+        self._test_create_valid_user()
+
+    def test_create_valid_global_admin(self):
+        self._test_create_valid_global_admin()
+
+    def test_create_user_duplicate_username(self):
+        first_user = self._test_create_valid_user()
+        self.assertIn('username', first_user)
+        username = first_user['username']
+
+        # generate second user info, but then use the same username as the first user
+        second_user_data = self.TestDataGen.gen_user()
+        second_user_data['username'] = username
+        second_user_add_params = self._convert_user_data_to_user_add_params(second_user_data)
+
+        # attempt create user with the same username
+        with self.assertRaises(requests.HTTPError) as cm:
+            cb.user_add_from_data(**second_user_add_params)
+
+        self.assertEqual(cm.exception.response.status_code, 409)
+
+    ## Delete user tests
+
+    def test_delete_unknown_user(self):
+        username = "unknown-user"
+
+        # verify the user does not exist on the server
+        self._assert_user_doesnt_exist(username)
+
+        # attempt to delete the user
+        with self.assertRaises(requests.HTTPError) as cm:
+            cb.user_del(username)
+
+        self.assertEqual(cm.exception.response.status_code, 500)
```

Comment:
I intended to put a TODO here as I did in `_assert_user_doesnt_exist` where I would've expected a different error code like a 400 or 404, but the API was sending back a 500.  

Also FYI, HTTP 409 is Conflict -- 403 is Forbidden.